### People's Names - Source Names Preprocessing

In [1]:
import re
import pandas as pd
import numpy as np

In [2]:
raw = pd.read_excel('People.xlsx', encoding='UTF-8-sig')
raw.head(10)

第一部分 外国姓名（日、朝、越等使用和曾使用汉字国家人名除外） Unnamed: 1 Unnamed: 2
0                               A        NaN        NaN
1                             Aab         瑞典         奥布
2                            Aaba          土         阿巴
3                           Aabel          挪        奥贝尔
4                         Aabraek          挪       奥布赖克
5                         Aabrant         瑞典       阿布兰特
6                            Aaby         瑞典         奥比
7                           Aabye     英、丹、瑞典         奥比
8                            Aach          德         阿赫
9                             Aad      荷、阿拉伯         阿德

In [3]:
raw.describe()

第一部分 外国姓名（日、朝、越等使用和曾使用汉字国家人名除外） Unnamed: 1 Unnamed: 2
count                           676872     672159     676700
unique                          641851      13668     501285
top                                San          法       穆罕默德
freq                                11      81876         41

In [5]:
raw = raw.rename({'第一部分 外国姓名（日、朝、越等使用和曾使用汉字国家人名除外）': 'source_name', 'Unnamed: 1': 'origin', 'Unnamed: 2':'target_name'}, axis=1)

In [8]:
raw = raw.dropna()

In [9]:
raw.describe()

source_name  origin target_name
count       672039  672039      672039
unique      637695   13656      496738
top            San       法        穆罕默德
freq            11   81876          41

字典里共有672039个词目

In [16]:
raw = raw.astype(str)

In [21]:
raw_en = raw[raw['origin'].str.contains('英')]

In [22]:
raw_en.describe()

source_name origin target_name
count        58456  58456       58456
unique       58396   2773       43472
top         Duarte      英          贝尔
freq             2  46767          18

In [55]:
for d in raw_en['source_name']:
    d = d.strip()
    if re.findall(re.compile('[^a-zA-Z]'), d):
        print(d)

A'Court
Adén
Alván
Andrásick
Bertram，James Munro
Bowes-Lyon
Brontё
Brzeziński
Castañeda
Crécy
Crèvecoeur
Csanádi
D<sup>′</sup>Abbes
D<sup>′</sup>Abernon
D<sup>′</sup>abreu
D<sup>′</sup>accone
D<sup>′</sup>acosta
D<sup>′</sup>Aeth
D<sup>′</sup>Agostini
D<sup>′</sup>Agostino
D<sup>′</sup>Aguilar
D<sup>′</sup>Albert
D<sup>′</sup>Albiac
D<sup>′</sup>Alessandro
D<sup>′</sup>alexander
D<sup>′</sup>alton
D<sup>′</sup>Amato
D<sup>′</sup>Ambrosio
D<sup>′</sup>Ambrumenil
D<sup>′</sup>Amico
D<sup>′</sup>Amours
D<sup>′</sup>Anastasio
D<sup>′</sup>Angelo
D<sup>′</sup>Aprano
D<sup>′</sup>Aquino
D<sup>′</sup>Arcy
D<sup>′</sup>Avigdor
D<sup>′</sup>Costa
D<sup>′</sup>Eauthreau
De Forest
D<sup>′</sup>egville
Del junco
Dell<sup>′</sup>Omo
Dell<sup>′</sup>Osso
Del renzio
Del riego
Del strother
Del tufo
Den Uyl
D<sup>′</sup>erlanger
Des Champs
Des Graz
Des Voeux
D<sup>′</sup>eyncourt
D<sup>′</sup>harnoncourt
D<sup>′</sup>Hooghe
D<sup>′</sup>iorio
D<sup>′</sup>itro
D<sup>′</sup>jamoos
Djordjević
D<sup>′</su

In [64]:
count = 0
for index, dp in raw_en.iterrows():
    src = dp['source_name']
    if re.findall(re.compile('[^a-zA-Z]'), src):
        print(src)
        src = src.replace(' ', '')
        src = src.replace('<sup>′</sup>', '\'')
        dp['source_name'] = src
        count += 1

A'Court
Adén
Alván
Andrásick
Bertram，James Munro
Bowes-Lyon
Brontё
Brzeziński
Castañeda
Crécy
Crèvecoeur
Csanádi
D<sup>′</sup>Abbes
D<sup>′</sup>Abernon
D<sup>′</sup>abreu
D<sup>′</sup>accone
D<sup>′</sup>acosta
D<sup>′</sup>Aeth
D<sup>′</sup>Agostini
D<sup>′</sup>Agostino
D<sup>′</sup>Aguilar
D<sup>′</sup>Albert
D<sup>′</sup>Albiac
D<sup>′</sup>Alessandro
D<sup>′</sup>alexander
D<sup>′</sup>alton
D<sup>′</sup>Amato
D<sup>′</sup>Ambrosio
D<sup>′</sup>Ambrumenil
D<sup>′</sup>Amico
D<sup>′</sup>Amours
D<sup>′</sup>Anastasio
D<sup>′</sup>Angelo
D<sup>′</sup>Aprano
D<sup>′</sup>Aquino
D<sup>′</sup>Arcy
D<sup>′</sup>Avigdor
D<sup>′</sup>Costa
D<sup>′</sup>Eauthreau
De Forest
D<sup>′</sup>egville
Del junco
Dell<sup>′</sup>Omo
Dell<sup>′</sup>Osso
Del renzio
Del riego
Del strother
Del tufo
Den Uyl
D<sup>′</sup>erlanger
Des Champs
Des Graz
Des Voeux
D<sup>′</sup>eyncourt
D<sup>′</sup>harnoncourt
D<sup>′</sup>Hooghe
D<sup>′</sup>iorio
D<sup>′</sup>itro
D<sup>′</sup>jamoos
Djordjević
D<sup>′</su

In [65]:
print(count)

352


352 srouce names containing non-English letters, keep those with spaces and replace $<sup>′</sup>$ with $'$.

In [74]:
count = 0
for index, dp in raw_en.iterrows():
    src = dp['source_name']
    if re.findall(re.compile('[^a-zA-Z]'), src):
        if not re.match(r'^[a-zA-Z\']+$', src):
            print(src)
            raw_en = raw_en.drop(index=index)
            count += 1

Adén
Alván
Andrásick
Bertram，JamesMunro
Bowes-Lyon
Brontё
Brzeziński
Castañeda
Crécy
Crèvecoeur
Csanádi
Djordjević
Duché
Duprée
Grimké
Guenièvre
Haidée
Ibargüen
Irenӓus
Klór
Kobbé
Legaré
Muñoz
O’melveny
Orde-Powlett
Peña
Pérault
Pérot
Péterffy
Planché
Prédmore
Raánan
Ramée
Schӧeffel
Schӧem
Schӧenbrun
Schӧeni
Schӧenmann
Schӧeppel
Schüerch
Schüessler
Schüettinger
Schůt
Şlesinger
St(Saint)
Strebél
Tourgée
Tourjée
Vokatý
Zídek
Zoё


In [75]:
print(count)

51


51 exceptional names needed to be removed.

In [77]:
raw_en.describe()

source_name origin target_name
count        58405  58405       58405
unique       58326   2772       43439
top            Don      英          贝尔
freq             2  46744          18

####  Make sure the first step of data extraction (source name) is done right.

In [94]:
count = 0
for index, dp in raw_en.iterrows():
    src = dp['source_name']
    if re.match(r'^[a-zA-Z\']+$', src):
        if '\'' in src:
            print(src)
            count += 1
    else:
        print('Error!', src)

A'Court
D'Abbes
D'Abernon
D'abreu
D'accone
D'acosta
D'Aeth
D'Agostini
D'Agostino
D'Aguilar
D'Albert
D'Albiac
D'Alessandro
D'alexander
D'alton
D'Amato
D'Ambrosio
D'Ambrumenil
D'Amico
D'Amours
D'Anastasio
D'Angelo
D'Aprano
D'Aquino
D'Arcy
D'Avigdor
D'Costa
D'Eauthreau
D'egville
Dell'Omo
Dell'Osso
D'erlanger
D'eyncourt
D'harnoncourt
D'Hooghe
D'iorio
D'itro
D'jamoos
D'olier
D'Orazio
D'oyly
D'Urfey
D'Ewes
I'Anson
L'Amour
L'Estrange
L'Hagan
Maco'shannaig
M'Gregor
O'Ballance
O'Beirne
O'Boyle
O'Brady
O'Brian
O'Brien
O'brine
O'Bryan
O'Byrne
O'Callaghan
O'Carolan
O'Casey
O'Cassy
O'Ceallaigh
O'Clery
O'Collins
O'Connell
O'Conner
O'Conners
O'Connor
O'Conor
O'Coraill
O'Curry
O'Dalaigh
O'Daniel
O'Day
O'Dea
O'Derrig
O'Doherty
O'Donnel
O'Donnell
O'Donoghue
O'Donohue
O'Donovan
O'Dowd
O'Dowda
O'Driscoll
O'Duffy
O'Dwyer
O'Faolain
O'Farrell
O'Ferrall
O'Fiaich
O'Flahertie
O'Flaherty
O'Flynn
O'Gallar
O'Gara
O'Geron
O'Gmole
O'Gorman
O'Grad
O'Grady
O'Graham
O'Hagan
O'Halloran
O'Hanlon
O'Hannrachain
O'Hara
O'Ha

In [95]:
print(count)

161


In [100]:
raw_en = raw_en.reset_index(drop=True)
raw_en.to_excel('people_src_processed.xlsx', index=False, encoding='UTF-8-sig')

In [104]:
raw_en = raw_en.drop(columns=['index'])

In [107]:
raw_en.to_excel('people_src_processed.xlsx', index=False, encoding='UTF-8-sig')

####  May be useful afterwards (directly processing the old version of people.txt)

In [ ]:
with open('People.txt', 'r', encoding='UTF-8-sig') as f:
    data = f.readlines()

In [ ]:
raw_pattern = r'(.+?)\t(.+?)\t(.+)\n'
raw_df = pd.DataFrame(columns=['English', 'Chinese'])
en_names = []
ch_names = []
sources = []

for line in data:
    pairs = re.findall(raw_pattern, line)
    if pairs:
        en_names.append(pairs[0][0])
        ch_names.append(pairs[0][2])
        sources.append(pairs[0][1])
    else:
        print(pairs)

In [ ]:
raw_df['English'] = en_names
raw_df['Chinese'] = ch_names
raw_df['Source'] = sources
raw_df.describe()

In [ ]:
raw_df[raw_df['English'] == 'Finola']

### Remove noise '(教名...)' 

In [ ]:
bracket_pattern = r'(.+)?\((.+)\)'
counts = dict()
temp = 0
temp1 = 0
temp2 = 0
for index, dp in raw_df.iterrows():
    ch = dp['Chinese']
    if re.findall(bracket_pattern, ch):
        temp += 1
        pairs = re.findall(bracket_pattern, ch)
        content_in_bracket = pairs[0][1]
        if '教名' in content_in_bracket:
            temp1 += 1
            dp['Chinese'] = pairs[0][0]
            # print(pairs[0][0])
            if not pairs[0][0] in raw_df['Chinese']:
                temp2 += 1
                print(dp['English'], dp['Chinese'], dp['Source'])
counts['first'] = [temp, temp1, temp2] 

In [ ]:
print(counts)
raw_df.describe()

去掉“教名”以后，unique Chinese names从43474减少到43144，共少了330个附加词“教名”。

In [ ]:
[t for t in raw_df['Source'] if not '英' in t]

In [ ]:
raw_df[(raw_df['Source'] == '美')]

In [ ]:
raw_df[(raw_df['Source'] == '意')]

### Remove noise '(女名...)' 

In [ ]:
temp = 0
for index, dp in raw_df.iterrows():
    ch = dp['Chinese']
    if re.findall(bracket_pattern, ch):
        temp += 1
        pairs = re.findall(bracket_pattern, ch)
        # print(dp['English'], pairs, dp['Source'])
        content_in_bracket = pairs[0][1]
        if '女名' in content_in_bracket:
            # 大部分情况只有女名一个标注
            if not content_in_bracket == '女名':
                print(content_in_bracket)
            dp['Chinese'] = pairs[0][0]
counts.append(temp)

In [ ]:
raw_df.describe()

去掉“教名”以后，unique Chinese names从43144减少到43133，共少了11个附加词“女名”。

In [ ]:
counts

In [ ]:
temp = 0
for index, dp in raw_df.iterrows():
    ch = dp['Chinese']
    if re.findall(bracket_pattern, ch):
        count1 += 1
        pairs = re.findall(bracket_pattern, ch)
        content_in_bracket = pairs[0][1]